FirebaseAuth/Sources/Swift/User/User.swift

```diff
@@ -0,0 +1,2076 @@
+      @param user The user to pass to callback if there is no error.
+      @param error The error to pass to callback.
+   */
+  private class func callInMainThreadWithUserAndError(callback: ((User?, Error?) -> Void)?,
+                                                      user: User,
+                                                      error: Error?) {
+    if let callback {
+      DispatchQueue.main.async {
+        callback((error != nil) ? nil : user, error)
+      }
+    }
+  }
+
+  /** @fn callInMainThreadWithAuthDataResultAndError
+      @brief Calls a callback in main thread with user and error.
+      @param callback The callback to be called in main thread.
+      @param result The result to pass to callback if there is no error.
+      @param error The error to pass to callback.
+   */
+  private class func callInMainThreadWithAuthDataResultAndError(callback: ((AuthDataResult?,
+                                                                            Error?) -> Void)?,
+  result: AuthDataResult?,
+  error: Error?) {
+    if let callback {
+      DispatchQueue.main.async {
+        callback(result, error)
+      }
+    }
+  }
+
+  // MARK: NSSecureCoding
+
+  private let kUserIDCodingKey = "userID"
+  private let kHasEmailPasswordCredentialCodingKey = "hasEmailPassword"
+  private let kAnonymousCodingKey = "anonymous"
+  private let kEmailCodingKey = "email"
+  private let kPhoneNumberCodingKey = "phoneNumber"
+  private let kEmailVerifiedCodingKey = "emailVerified"
+  private let kDisplayNameCodingKey = "displayName"
+  private let kPhotoURLCodingKey = "photoURL"
+  private let kProviderDataKey = "providerData"
+  private let kAPIKeyCodingKey = "APIKey"
+  private let kFirebaseAppIDCodingKey = "firebaseAppID"
+  private let kTokenServiceCodingKey = "tokenService"
+  private let kMetadataCodingKey = "metadata"
+  private let kMultiFactorCodingKey = "multiFactor"
+  private let kTenantIDCodingKey = "tenantID"
+
+  public static var supportsSecureCoding = true
+
+  public func encode(with coder: NSCoder) {
+    coder.encode(uid, forKey: kUserIDCodingKey)
+    coder.encode(isAnonymous, forKey: kAnonymousCodingKey)
+    coder.encode(hasEmailPasswordCredential, forKey: kHasEmailPasswordCredentialCodingKey)
+    coder.encode(providerData, forKey: kProviderDataKey)
+    coder.encode(email, forKey: kEmailCodingKey)
+    coder.encode(phoneNumber, forKey: kPhoneNumberCodingKey)
+    coder.encode(isEmailVerified, forKey: kEmailVerifiedCodingKey)
+    coder.encode(photoURL, forKey: kPhotoURLCodingKey)
+    coder.encode(displayName, forKey: kDisplayNameCodingKey)
+    coder.encode(metadata, forKey: kMetadataCodingKey)
+    coder.encode(tenantID, forKey: kTenantIDCodingKey)
+    if let auth {
+      coder.encode(auth.requestConfiguration.APIKey, forKey: kAPIKeyCodingKey)
+      coder.encode(auth.requestConfiguration.appID, forKey: kFirebaseAppIDCodingKey)
+    }
+    coder.encode(tokenService, forKey: kTokenServiceCodingKey)
+    #if os(iOS)
```

Comment:
thinking aloud for future improvements, maybe codable could help cut down this boilerplate? Auth serializes a couple pieces of data. And I bet we can handle the platform specific properties by #if'ing in their definitions on acceptable platforms